### PR TITLE
winePackages{unstable,staging}: 9.0 -> 9.2, cleanup

### PR DESCRIPTION
**Pull request by SuperSamus**
## Description of changes

https://gitlab.winehq.org/wine/wine/-/releases/wine-9.1
https://gitlab.winehq.org/wine/wine/-/releases/wine-9.2
The explanations for the other changes are in the commit messages.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by reckenrode** [comment]
LGTM. Running nixpkgs-review on Darwin now.

I’m working on some changes to `overrideSDK` on Darwin that should let me remove some of the workarounds and drop the patches. I also have patches for MetalD3D compatibility, which depends on the `overrideSDK` changes.

File: pkgs/applications/emulators/wine/base.nix
Comment (reckenrode) [inline pkgs/applications/emulators/wine/base.nix:213]: Did you mean to remove the `platforms` definition? Wayland isn’t supported on Darwin.

Comment (SuperSamus) [reply pkgs/applications/emulators/wine/base.nix:213]: It is possible to build `wine-wayland` without Wayland. (This line was originally checking for the "wayland" release)

If you meant "the evaluation will fail if `waylandSupport` is true because the dependencies don't exist on Darwin"... you are right, but that would also apply to other dependencies too.

Comment (SuperSamus) [reply pkgs/applications/emulators/wine/base.nix:213]: So, options in my mind:
- In `base.nix`, don't just check for `dependencySupport`, but check for `dependencySupport && stdenv.isRightOS`, so that the flag is ignored on the wrong operating system
  - This however causes the problem of having stuff like `wine-wayland` without Wayland on Darwin.
- In `base.nix`, add a list of `supportFlags` that aren't on Darwin: if any of these are true, disable Darwin support (which is the approach of before, but not just for Wayland)
  - This may make a Wine build accidentally incompatible with Darwin, and it might be hard to understand why. In `wine-packages.nix` you have to be careful on what is `= stdenv.isLinux`.

**Comment by reckenrode** [comment]
Result of `nixpkgs-review pr 286580` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>4 packages failed to build:</summary>
  <ul>
    <li>wine64Packages.wayland</li>
    <li>wine64Packages.waylandFull</li>
    <li>wineWow64Packages.wayland</li>
    <li>wineWow64Packages.waylandFull</li>
  </ul>
</details>
<details>
  <summary>14 packages built:</summary>
  <ul>
    <li>wine64 (wine64Packages.full ,wine64Packages.stableFull)</li>
    <li>wine64Packages.base (wine64Packages.stable)</li>
    <li>wine64Packages.minimal</li>
    <li>wine64Packages.staging</li>
    <li>wine64Packages.stagingFull</li>
    <li>wine64Packages.unstable</li>
    <li>wine64Packages.unstableFull</li>
    <li>wineWow64Packages.base (wineWow64Packages.stable)</li>
    <li>wineWow64Packages.full (wineWow64Packages.stableFull)</li>
    <li>wineWow64Packages.minimal</li>
    <li>wineWow64Packages.staging</li>
    <li>wineWow64Packages.stagingFull</li>
    <li>wineWow64Packages.unstable</li>
    <li>wineWow64Packages.unstableFull</li>
  </ul>
</details>

File: pkgs/applications/emulators/wine/base.nix
Comment (reckenrode) [reply pkgs/applications/emulators/wine/base.nix:213]: The wine-wayland packages fail to build on Darwin. I’d like to avoid introducing unnecessary Hydra failures.

> So, options in my mind:
> 
> * In `base.nix`, don't just check for `dependencySupport`, but check for `dependencySupport && stdenv.isRightOS`, so that the flag is ignored on the wrong operating system
>   
>   * This however causes the problem of having stuff like `wine-wayland` without Wayland on Darwin.
> * In `base.nix`, add a list of `supportFlags` that aren't on Darwin: if any of these are true, disable Darwin support (which is the approach of before, but not just for Wayland)

Silently disabling requested support for a feature would be a bad experience for users. It would be better add Darwin to `badPlatforms` for flags known to fail to build on that platform.

>   * This may make a Wine build accidentally incompatible with Darwin, and it might be hard to understand why. In `wine-packages.nix` you have to be careful on what is `= stdenv.isLinux`.

Some of the flags set in `wine-packages.nix` may result in working builds on Darwin but don’t make sense being enabled by default (such as `gtkSupport` and `x11Support`). If a user really wants to enable one of them anyway, they can override one of the Wine packages to enable it.

**Comment by SuperSamus** [comment]
Alright, now `wine-wayland` should be unsupported on Darwin like it was before.
Also, `x11Support` and `waylandSupport` are now set in `wine-packages.nix`.
(Note: both of them are false in `minimal`. Should `x11Support` be true even on `minimal`?)

**Comment by reckenrode** [comment]
I’m pretty sure that’s way too aggressive for marking things broken, but I won’t be able to dig into it until later this evening. 

The reason for marking Wayland support broken on Darwin is because the `wayland` derivation builds, so Wine tries to build with support and fails.

I’ll defer to the Linux maintainers whether `x11Support` should be enabled in the minimal build on Linux. It may be that it’s meant for headless use cases.

(This exchange makes me wonder if Wine should be building with libpcap on Darwin. Upstream’s build does. I can address that in my Darwin cleanup PR.)

**Comment by SuperSamus** [comment]
IIRC (I don't remember where I read this), the Collabora fork still has some advantages over regular Wine with Wayland (e.g., some games are perfectly playable, which isn't the case on regular Wine).
Don't know how valid this is today.

Also, I'd rather remove rather than alias to something else.
(Though that can be done in another PR.)

**Comment by reckenrode** [comment]
My apologies for not getting back sooner re: Darwin supported flags. https://github.com/NixOS/nixpkgs/pull/287609 ended up taking more time than expected. I looked back over the `darwinUnsupportedFlags`, and I think the set of flags is fine. Core Text is used on Darwin instead of fontconfig. x11 support should work, but it doesn’t make a lot of sense. The others are all Linux-specific (or not relevant to Darwin).

I’ll take care of libpcap plus the other Darwin cleanups once I’m properly able to specify the SDK version and deployment target. Those cleanups will also include the patches needed to support https://github.com/NixOS/nixpkgs/issues/236414.

The only other question I have is whether this should be bumped to Wine 9.2, which just came out a few hours ago.

**Comment by SuperSamus** [comment]
`wine-staging` still doesn't have a release tagged 9.2, I guess it will tomorrow?
(Speaking of staging, shouldn't it be migrated to `fetchFromGitLab`, since its GitHub repo is a mirror now?)

**Comment by 7c6f434c** [comment]
Switching to the true-primary sounds good to me.

**Review by reckenrode — APPROVED** [review]
I agree about switching wine-staging to the upstream GitLab instance. Changes lgtm.

**Pull request merged**
